### PR TITLE
JAMES-2188 A user can't set a message in a mailbox from someone else.

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java [hunk 1]
```diff
@@ -128,6 +137,7 @@ private void handleCreate(CreationMessageEntry create, Builder responseBuilder,
             validateImplementedFeature(create, mailboxSession);
             validateArguments(create, mailboxSession);
             validateRights(create, mailboxSession);
+            validateIsUserOwnerOfMailboxes(create, mailboxSession);
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java]: why it's not covered by validateRights ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java [hunk 2]
```diff
@@ -239,7 +257,20 @@ private boolean isAllowedFromAddress(CreationMessage creationMessage, List<Strin
...
+            throw new MailboxRightsException();
+        }
+    }
+
+    private boolean containsAnyDelegatedMailbox(List<String> mailboxIds, MailboxSession session) {
+        return mailboxIds.stream()
+            .map(mailboxIdFactory::fromString)
+            .map(Throwing.function(mailboxId -> mailboxManager.getMailbox(mailboxId, session)))
+            .map(Throwing.function(MessageManager::getMailboxPath))
+            .anyMatch(path -> !path.getUser().equals(session.getUser().getUserName()));
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java]: there is a better comparator function for username

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java [hunk 3]
```diff
@@ -239,7 +257,20 @@ private boolean isAllowedFromAddress(CreationMessage creationMessage, List<Strin
...
     }
 
-    
+    @VisibleForTesting void validateIsUserOwnerOfMailboxes(CreationMessageEntry entry, MailboxSession session) throws MailboxRightsException {
+        if (containsAnyDelegatedMailbox(entry.getValue().getMailboxIds(), session)) {
+            throw new MailboxRightsException();
+        }
+    }
+
+    private boolean containsAnyDelegatedMailbox(List<String> mailboxIds, MailboxSession session) {
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java]: it seems this function is misnamed as we don't mind if the mailbox is delegated, but only if the user is not the owner of one the mailboxes

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java (hunk 1 above)
Comment (aduprat) [reply server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMessagesCreationProcessor.java]: validateRights is just checking that one of the sender is the current user

**Comment by mbaechler** [comment]
test this please

**Comment by mbaechler** [comment]
test this please

**Comment by mbaechler** [comment]
test this please

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesCreationProcessorTest.java [hunk 4]
```diff
@@ -133,13 +140,20 @@ public void setUp() throws MailboxException {
         mockedMailSpool = mock(MailSpool.class);
         mockedMailFactory = mock(MailFactory.class);
         mockedAttachmentManager = mock(AttachmentManager.class);
+        mockedMailboxManager = mock(MailboxManager.class);
```
Comment (chibenwa) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesCreationProcessorTest.java]: Why not use the in memory one? Won't it be easier?

**Comment by aduprat** [comment]
test this please

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesCreationProcessorTest.java (hunk 4 above)
Comment (aduprat) [reply server/protocols/jmap/src/test/java/org/apache/james/jmap/methods/SetMessagesCreationProcessorTest.java]: I prefere to keep the mock behaviour of this test.

**Comment by aduprat** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**